feat: optimize solidity honk verifier

Maddiaa0 opened the pull request:
Complete, without zk

Cuts off a million 

next steps:
proof size has a large consequence with this one, due to memory expansion, so having a different proof type for the keccak one which does tighter serialisation will save a good chunk of gas - working on this next
- pr here https://github.com/AztecProtocol/aztec-packages/pull/16018

Maddiaa0 commented:
* **#8690** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/8690?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Maddiaa0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

TomAFrench commented on barretenberg/sol/src/honk/Relations.sol:
```diff
@@ -633,13 +633,15 @@ library RelationsLib {
 
-        evals[20] = evals[20] + ep.q_pos_by_scaling * (ep.v3 - wire(p, WIRE.W_O_SHIFT));
+        evals[21] = ep.q_pos_by_scaling * (ep.v4 - wire(p, WIRE.W_4_SHIFT));
 
-        evals[21] = evals[21] + ep.q_pos_by_scaling * (ep.v4 - wire(p, WIRE.W_4_SHIFT));
+        evals[21] = ep.q_pos_by_scaling * (ep.v4 - wire(p, WIRE.W_4_SHIFT));
```
Looks accidentally duplicated.

lucasxia01 commented on barretenberg/acir_tests/sol-test/src/index.js:
```diff
@@ -294,6 +294,9 @@ try {
         console.log("Deploying ZKTranscriptLib library...");
         const libraryAddress = await deploy(signer, libraryAbi, libraryBytecode);
+
+        // Wait for the library deployment - for some reason we have an issue with nonces here
```
what do we do about this

Maddiaa0 replied:
the sleep here is enough for the meantime, another option would be manually assigning nonced

lucasxia01 commented on barretenberg/cpp/src/barretenberg/dsl/acir_proofs/honk_contract.hpp:
```diff
@@ -1054,7 +1055,7 @@ library RelationsLib {
         ap.record_delta = wire(p, WIRE.W_4_SHIFT) - wire(p, WIRE.W_4);
 
-        ap.index_is_monotonically_increasing = ap.index_delta * ap.index_delta - ap.index_delta; // deg 2
+        ap.index_is_monotonically_increasing = ap.index_delta * (ap.index_delta - Fr.wrap(1)); // deg 2
```
is this cheaper because the value we're subtracting is smaller?

lucasxia01 commented on barretenberg/sol/src/honk/keys/EcdsaHonkVerificationKey.sol:
```diff
@@ -95,8 +95,8 @@ library EcdsaHonkVerificationKey {
                 y: uint256(0x0cd29f3121acf9430707827d9b0805f991402d944261e1d648d9c08c7cec5475)
             }),
             t3: Honk.G1Point({
-                x: uint256(0x1df7f08d004e38c6cc24155081bf68c1a6444b526bd98beea00feabc8ea337f9),
```
why are these changing?

Maddiaa0 commented on barretenberg/acir_tests/sol-test/HonkTest.sol:
```diff
@@ -12,7 +12,7 @@ contract Test {
     }
 
-    function test(bytes calldata proof, bytes32[] calldata publicInputs) public view returns (bool) {
+    function test(bytes calldata proof, bytes32[] calldata publicInputs) public returns (bool) {
```
it actually should be one again! As it doesnt update state view

ludamad commented on barretenberg/sol/bootstrap.sh:
```diff
@@ -34,7 +34,7 @@ function build_sol {
         # Ensure libraries are at the correct version
         git submodule update --init --recursive ./lib
```
was this temporary?

Maddiaa0 replied:
it was added when this pr was first opened as the formatter could not handle the asm blocks. I think it may be fixed with newer foundries

ludamad commented on barretenberg/sol/scripts/copy_optimized_to_cpp.sh:
```diff
@@ -0,0 +1,316 @@
+#!/bin/bash
+
+# Script to copy optimized Solidity verifier files into the C++ honk_optimized_contract.hpp file
+# This automates the manual process of copying optimized verifier contracts
```
Comments from slack:

No comments are blocking, ok with this going in as-is for now

but - we can probably remove a bit of brittleness. you're doing template discovery on what would normally be an output artifact, then doing find replace with strings.
What about:
```
the script copies the solidity as-is. but the solidity has e.g. // c++-replace VK_HASH comments.
the c++-replace directives can affect the next line
2. store the C++ artifacts in a map of symbol to string
3. process the c++-replace directives
probably error on any unused symbols
```

ludamad replied:
the other thing is, it's more normal to bundle artifacts in C++ as its own header file

AztecBot committed:
feat: optimize solidity honk verifier

Complete, without zk

Cuts off a million

next steps:
proof size has a large consequence with this one, due to memory expansion, so having a different proof type for the keccak one which does tighter serialisation will save a good chunk of gas - working on this next
- pr here https://github.com/AztecProtocol/aztec-packages/pull/16018
